components/homme/src/share/gllfvremap_mod.F90

```diff
@@ -0,0 +1,2253 @@
+     module procedure gfr_fv_phys_to_dyn_dom_mt
+  end interface gfr_fv_phys_to_dyn
+
+  interface gfr_fv_phys_to_dyn_topo
+     module procedure gfr_fv_phys_to_dyn_topo_hybrid
+     module procedure gfr_fv_phys_to_dyn_topo_dom_mt
+  end interface gfr_fv_phys_to_dyn_topo
+
+  interface gfr_pg1_reconstruct
+     module procedure gfr_pg1_reconstruct_hybrid
+     module procedure gfr_pg1_reconstruct_dom_mt
+  end interface gfr_pg1_reconstruct
+
+  interface gfr_pg1_reconstruct_topo
+     module procedure gfr_pg1_reconstruct_topo_hybrid
+     module procedure gfr_pg1_reconstruct_topo_dom_mt
+  end interface gfr_pg1_reconstruct_topo
+
+contains
+
+  ! ----------------------------------------------------------------------
+  ! Public API.
+
+  subroutine gfr_init(par, elem, nphys, check, boost_pg1)
+    ! Initialize the gfr internal data structure.
+    !   nphys is N in pgN.
+    !   check is optional and defaults to false. It will produce very
+    ! verbose output if something goes wrong. It is also expensive. It
+    ! is intended to be used in unit testing and (if ever needed) for
+    ! debuggin.
+
+    use kinds, only: iulog
+    use dimensions_mod, only: nlev
+    use parallel_mod, only: parallel_t, abortmp
+    use quadrature_mod, only : gausslobatto, quadrature_t
+    use control_mod, only: ftype
+
+    type (parallel_t), intent(in) :: par
+    type (element_t), intent(in) :: elem(:)
+    integer, intent(in) :: nphys
+    logical, intent(in), optional :: check, boost_pg1
+
+    real(real_kind) :: R(npsq,nphys_max*nphys_max), tau(npsq)
+    integer :: nphys2
+
+    gfr%check = .false.
+    if (present(check)) gfr%check = check
+
+    gfr%boost_pg1 = .false.
+    if (present(boost_pg1)) gfr%boost_pg1 = boost_pg1    
+
+    gfr%tolfac = one
+    if (par%masterproc) then
+       write(iulog, '(a,i3,a,l2,a,i2,a,l2)') 'gfr> init nphys', nphys, ' check', gfr%check, &
+            ' ftype', ftype, ' boost_pg1', gfr%boost_pg1
+       if (nphys == 1) then
+          ! Document state of pg1. dcmip2016_test1 shows it is too coarse. For
+          ! boost_pg1 = true, stepon's DSS loop needs to be separated from its
+          ! tendency application loop.
+          write(iulog,*), 'gfr> Warning: pg1 is too coarse; see comments at top of gllfvremap_mod.F90'
+          if (.not. gfr%boost_pg1) then
+             write(iulog,*), 'gfr> Warning: If you want to try pg1, use the boosted-accuracy &
+                  boost_pg1 option and call gfr_pg1_reconstruct(_topo).'
+          end if
+       end if
+    end if
+
+    if (nphys > np) then
+       ! The FV -> GLL map is defined only if nphys <= np. If we ever are
+       ! interested in the case of nphys > np, we will need to write a different
+       ! map. See "!assume" annotations for mathematical assumptions in
+       ! particular routines.
+       call abortmp('gllfvremap_mod: nphys must be <= np')
+    end if
+    if (qsize == 0) then
+       call abortmp('gllfvremap_mod: qsize must be >= 1')
```

Comment:
Why is qsize>0 needed?

Reply:
Well, I suppose it isn't. But a use case for physgrid without any tracers seems contrived. Since I don't have a test for qsize=0, I prefer to abort on this b/c I think it's more useful to clearly say that an anomalous situation has arisen.